e2e/callback_test.go

```diff
@@ -0,0 +1,372 @@
+	require.NoError(suite.T(), err)
+
+	if relay {
+		suite.coordinator.UpdateTime()
+		suite.coordinator.RelayAndAckPendingPackets(suite.path)
+		suite.coordinator.UpdateTime()
+	}
+}
+
+func sendIcsFromChainB(suite *CbTestSuite, nft, token_id, memo string, relay bool) {
+	msg := fmt.Sprintf(`{ "send_nft": {"cw721": "%s", "ics721": "%s", "token_id": "%s", "recipient":"%s", "channel_id":"%s", "memo":"%s"}}`, nft, suite.bridgeB.String(), token_id, suite.testerA.String(), suite.path.EndpointB.ChannelID, memo)
+	_, err := suite.chainB.SendMsgs(&wasmtypes.MsgExecuteContract{
+		Sender:   suite.chainB.SenderAccount.GetAddress().String(),
+		Contract: suite.testerB.String(),
+		Msg:      []byte(msg),
+		Funds:    []sdk.Coin{},
+	})
+	require.NoError(suite.T(), err)
+
+	if relay {
+		suite.coordinator.UpdateTime()
+		suite.coordinator.RelayAndAckPendingPackets(suite.path.Invert())
+	}
+}
+
+func queryTesterSent(t *testing.T, chain *wasmibctesting.TestChain, tester string) string {
+	resp := TesterResponse{}
+	testerSentQuery := TesterSentQuery{
+		GetSentCallback: EmptyData{},
+	}
+	err := chain.SmartQuery(tester, testerSentQuery, &resp)
+	require.NoError(t, err)
+	return *resp.Owner
+}
+
+func queryTesterReceived(t *testing.T, chain *wasmibctesting.TestChain, tester string) string {
+	resp := TesterResponse{}
+	testerReceivedQuery := TesterReceivedQuery{
+		GetReceivedCallback: EmptyData{},
+	}
+	err := chain.SmartQuery(tester, testerReceivedQuery, &resp)
+	require.NoError(t, err)
+	return *resp.Owner
+}
+
+func queryTesterReceivedNftContract(t *testing.T, chain *wasmibctesting.TestChain, tester string) string {
+	resp := ""
+	testerReceivedQuery := TesterReceivedNftContractQuery{
+		GetReceivedNftContract: EmptyData{},
+	}
+	err := chain.SmartQuery(tester, testerReceivedQuery, &resp)
+	require.NoError(t, err)
+	return resp
+}
+
+func queryTesterReceivedErr(t *testing.T, chain *wasmibctesting.TestChain, tester string) error {
+	resp := TesterResponse{}
+	testerReceivedQuery := TesterReceivedQuery{
+		GetReceivedCallback: EmptyData{},
+	}
+	err := chain.SmartQuery(tester, testerReceivedQuery, &resp)
+	return err
+}
+
+func (suite *CbTestSuite) TestSuccessfulTransfer() {
+	memo := callbackMemo(nftSentCb(), "", nftReceivedCb(), "")
+	sendIcsFromChainA(suite, suite.cw721A.String(), "2", memo, true)
+
+	// Query the owner of NFT on cw721
+	chainAOwner := queryGetOwnerOf(suite.T(), suite.chainA, suite.cw721A.String(), "2")
+	require.Equal(suite.T(), chainAOwner, suite.bridgeA.String())
+	chainBOwner := queryGetOwnerOf(suite.T(), suite.chainB, suite.cw721B.String(), "2")
+	require.Equal(suite.T(), chainBOwner, suite.testerB.String())
+
+	// We query the data we have on the tester contract
+	// This ensures that the callbacks are called after all the messages was completed
+	// and the transfer was successful
+	testerDataOwnerA := queryTesterSent(suite.T(), suite.chainA, suite.testerA.String())
+	require.Equal(suite.T(), testerDataOwnerA, suite.bridgeA.String())
```

Comment:
Ok, this is good. Here we check NFT on chain A is escroweb by ics721 (bridge A).